httpcore/_async/connection.py

```diff
@@ -81,13 +81,15 @@ async def handle_async_request(self, request: Request) -> Response:
                             stream=stream,
                             keepalive_expiry=self._keepalive_expiry,
                         )
-                except Exception as exc:
-                    self._connect_failed = True
-                    raise exc
-            elif not self._connection.is_available():
-                raise ConnectionNotAvailable()
+                elif not self._connection.is_available():
+                    raise ConnectionNotAvailable()
 
-        return await self._connection.handle_async_request(request)
+            return await self._connection.handle_async_request(request)
+        except BaseException as exc:
+            if not isinstance(exc, ConnectionNotAvailable):
+                self._connect_failed = True
+
+            raise exc
```

Comment:
I believe using `raise exc` will add extra frames to the traceback
```suggestion
            raise
```